### PR TITLE
RtVectorLayerService

### PR DESCRIPTION
**Pull request by thiloSchlemmer**
add a VectorlayerService to handle realtime features

### CODE REVIEW

File: src/modules/olMap/services/RtVectorLayerService.js [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+	}
+
+	/**
+	 * Builds an ol VectorLayer from an VectorGeoResource
+	 * @param {string} id layerId
+	 * @param {RtVectorGeoResource} rtVectorGeoResource
+	 * @param {OlMap} olMap
+	 * @returns olVectorLayer
+	 */
+	createVectorLayer(id, rtVectorGeoResource, olMap) {
```
Comment (taulinger) [inline src/modules/olMap/services/RtVectorLayerService.js:196]: Let's name it `createLayer` (see #2283)

File: src/modules/olMap/services/RtVectorLayerService.js [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+			opacity: opacity,
+			minZoom: minZoom ?? undefined,
+			maxZoom: maxZoom ?? undefined
+		});
+		const vectorSource = new VectorSource();
+		vectorLayer.setSource(vectorSource);
+
+		this._startWebSocket(rtVectorGeoResource, vectorLayer, olMap);
+
+		// HINT: currently is no support for clustered realtime vector data needed or planned for future releases
```
Comment (taulinger) [inline src/modules/olMap/services/RtVectorLayerService.js:148]: Outdated comment?

File: src/modules/olMap/services/RtVectorLayerService.js [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+		const tryNextPort = (port) => {
+			webSocket.onmessage = undefined;
+			const nextPort = getNextPort(WebSocket_Ports, port);
+			this._startWebSocket(rtVectorGeoResource, olVectorLayer, olMap, nextPort);
+		};
+
+		const failure = () => {
+			throw new UnavailableGeoResourceError('Realtime-data cannot be displayed for technical reasons.', rtVectorGeoResource.id);
+		};
```
Comment (taulinger) [inline src/modules/olMap/services/RtVectorLayerService.js:113]: Is this the untestable part of the code?

File: test/modules/olMap/services/RtVectorLayerService.test.js [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+		sanitizeStyles: () => {}
+	};
+
+	// beforeEach(() => {
+	// 	$injector.registerSingleton('MapService', mapService);
+	// });
+
+	// afterEach(() => {
+	// 	$injector.reset();
+	// });
```
Comment (taulinger) [inline test/modules/olMap/services/RtVectorLayerService.test.js:25]: Needed?

**Review by taulinger — COMMENTED** [review]
Nice!
Now we should wait until we have the test endpoint and then we have to implement some kind of zooming to the extent of the current feature

File: src/modules/olMap/services/RtVectorLayerService.js (hunk 3 above)
Comment (thiloSchlemmer) [reply src/modules/olMap/services/RtVectorLayerService.js:113]: yes

Comment (thiloSchlemmer) [reply src/modules/olMap/services/RtVectorLayerService.js]: now refactored, to be testable

**Review by taulinger — COMMENTED** [review]
Looks good! Let's implement some kind of zooming to the extent of the current feature now

File: src/modules/olMap/services/RtVectorLayerService.js [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+	}
+
+	_processMessage(messageData, olVectorLayer, featureReader) {
+		const olVectorSource = olVectorLayer.getSource();
+		olVectorSource.clear();
+		const features = featureReader(messageData);
+		olVectorSource.addFeatures(features);
+	}
+
+	_fitViewOptionally(olVectorLayer, olMap, firstFit) {
```
Comment (taulinger) [inline src/modules/olMap/services/RtVectorLayerService.js:97]: Maybe we should add an explanation about our fit-to-extend/change-center approach

**Review by taulinger — COMMENTED** [review]
Well done!

**Review by taulinger — COMMENTED** [review]
We should stop receiving events when the layer is set to invisible and start receiving again when set to visible.

A possible approach is monitoring the layer's `change:visible` event.

Currently, we do not close the WebSocket anyway!!

grafik

**Review by taulinger — APPROVED** [review]
Well done!

**Pull request merged**